File: built-in plugins dsp/mb-restore-nr.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbRestoreNrPlugin: PluginDefinition = { id: 'mb-restore-nr', slug: 'mb-restore-nr', name: 'MB Noise Reduction', category: 'effect', type: 'gate' as any, version: '1.0.0', description: 'Spectral noise reduction with noise profile learning', author: 'Max Booster', parameters: [{ id: 'reduction', name: 'Reduction', type: 'float', defaultValue: 20, minValue: 0, maxValue: 60, automatable: true }, { id: 'sensitivity', name: 'Sensitivity', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'smoothing', name: 'Smoothing', type: 'float', defaultValue: 5, minValue: 0, maxValue: 20, automatable: true }, { id: 'attack', name: 'Attack', type: 'float', defaultValue: 5, minValue: 0.5, maxValue: 50, automatable: true }], defaultPreset: { reduction: 20, sensitivity: 0.5, smoothing: 5, attack: 5 } };

export default MbRestoreNrPlugin;

File: built-in plugins dsp/mb-restore-nr.cpp
/**
 * MB Noise Reduction
 * Category : effect
 * Type     : gate
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Spectral noise reduction with noise profile learning
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_RESTORE_NR_H
#define MB_RESTORE_NR_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbRestoreNr : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-restore-nr";
    static constexpr const char* PLUGIN_NAME    = "MB Noise Reduction";
    static constexpr const char* PLUGIN_TYPE    = "gate";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float reduction = 20f;  // range [0, 60]
    float sensitivity = 0.5f;  // range [0, 1]
    float smoothing = 5f;  // range [0, 20]
    float attack = 5f;  // range [0.5, 50]
    };

    MbRestoreNr() = default;
    ~MbRestoreNr() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.reduction = std::clamp(params.reduction, 0f, 60f);
        params.sensitivity = std::clamp(params.sensitivity, 0f, 1f);
        params.smoothing = std::clamp(params.smoothing, 0f, 20f);
        params.attack = std::clamp(params.attack, 0.5f, 50f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Noise Reduction
        return input;
    }
};

#endif // MB_RESTORE_NR_H
